### PR TITLE
🔧 Use make to build the extension

### LINKED ISSUE #129: Switch build system to `Makefile` instead of scripts
### Discussed in https://github.com/Schneegans/Fly-Pie/discussions/128

<div type='discussions-op-text'>

<sup>Originally posted by **daPhipz** September 26, 2021</sup>
I recently dug into learning `make`, and now want to apply the things that I learned in the "real world".
AFAIK, the building of Fly-Pie happens through the `scripts/create-release.sh` script as of now - right?
While this eventually will build the zip, `make` has the advantage of being able to only re-build the things that have changed since the last build (and therefore taking considerably less time to finish) - but I'm sure you know that already :wink:

Are you open to changing the build system to `make`? As I said, I would happily do the work of porting the existing one since I want to practice a little bit ;)</div>

### PR DESCRIPTION
**Pull request by daPhipz**
Fixes #129.

I'll explain here what the different targets in the new `Makefile` do. In everyday development, you'll only need the `PHONY` targets, everything else is behind the scenes stuff.

### General `Makefile` syntax

- The `@` sign in front of a command stops it from being printed to stdout by `make`, leaving only the actual output of the command.
- `$@` refers to the target of a rule.
- `$<` refers to the first prerequisite of a rule.
- `$$` escapes the second dollar sign for `make` and passes it on to bash.
- `%` is like `*` in `bash`.

### PHONY targets

These are meant to be run by the developer.

#### `build`
Build the extension. Since it is the first target in the file, running `make` and `make build` do  the same thing.

#### `release`
Build the extension and additionally run pre-release checks.

#### `install`
Build the extension and install it to the system.

#### `uninstall`
Uninstall the extension. Duh.

#### `all-po`
Update all `po` files with the latest strings. Handles potentially necessary `pot` file updates automatically, before actually updating the `po` files.
I am not sure if this is necessary/needs changes, since translations are managed by Weblate now.

### Regular targets

These are meant to be executed automatically, not by the developer. They can of course be executed manually if needed.

#### `flypie@schneegans.github.com.zip`
If any of the prerequisites (or prerequisites for prerequisites for prerequisites etc etc) changed, build the zip.
I also added the capability to modify the value of `version` in `metadata.json` which lets you do this:

```
make release VERSION=XX    # or make / make build / make install
```

#### `resources/flypie.gresource`
If `resources/flypie.gresource.xml` changed, update `resources/flypie.gresource`.
I added the `--sourcedir="resources"` flag, because `make` gets executed with the project root as working directory. `cd`-ing would produce less easy-to-understand code than adding this flag.

#### `schemas/gschemas.compiled`
If `schemas/org.gnome.shell.extensions.flypie.gschema.xml` changed, update `schemas/gschemas.compiled`.

#### `locale/%/LC_MESSAGES/flypie.mo`
If a `po` file or `messages.pot` changed, compile the new translations.

#### `po/%.po`
If `po/flypie.pot` changed, update `%.po`.

#### `po/flypie.pot`
If any `*.js` or `*.ui` files changed, update the `pot` file with the latest strings.

#### `clean`
Remove all build files and leave the raw source code behind.
Did I miss any files here?

## TODO
There is still some work left to do outside the `Makefile` - but since this was primarily intended for a bit of `make` practice, I wanted to hear your opinion on this before doing any more work.

- [x] Did anything change in the translation workflow with the transition to Weblate?
- [x] Did I miss any files in `make clean`?
- [x] Update GitHub Workflows
- [ ] Update documentation
  - [x] [`docs/continuous-integration.md`](https://github.com/Schneegans/Fly-Pie/blob/develop/docs/continuous-integration.md)
  - [x] [`docs/installation.md`](https://github.com/Schneegans/Fly-Pie/blob/develop/docs/installation.md)
  - [x] [`docs/translating.md`](https://github.com/Schneegans/Fly-Pie/blob/develop/docs/translating.md)
  - [x] [`docs/release-management.md`](https://github.com/Schneegans/Fly-Pie/blob/develop/docs/release-management.md)
  - [ ] Any other places?

Thanks for having a look at it!

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,81 @@
+SHELL := /bin/bash
+
+JS_FILES = $(shell find -type f -and \( -name "*.js" \))
+UI_FILES = $(shell find -type f -and \( -name "*.ui" \))
```
Comment (daPhipz) [inline Makefile:4]: These variables contain all the `js`/`ui` files in the project.
Not really relevant in this project, but nice to know:

Variable assignments made with `:=` are "fixed" - meaning they are evaluated at creation and then saved as a hardcoded string.
Variable assignments made with `=` are variable - meaning they are evaluated each time they get called in the `Makefile`.
This means: Would, during the execution of this `Makefile`, another `ui` file be created and then `UI_FILES` gets called again, the new file would be included. Does this explanation make any sense? :exploding_head:

Comment (Schneegans) [reply Makefile:4]: Thanks for the explanation! Also the "General Makefile Syntax" at the top is very helpful!

**Comment by Schneegans** [comment]
This looks pretty clean already! Here are some thoughts:

* Regarding Weblate: The *.po files are updated by weblate, while the *.pot file needs to be updated manually. So there should be a PHONY target to only update the *.pot file. If we go for weblate-only we could drop the *.po targets at all. 
* I haven't tested it yet, but do I understand this correctly: The *.pot file will now be updated whenever `make` is executed? If so, I am not sure if that is a good thing. On the one hand, it will ensure that the source code positions mentioned in the file are always correct, on the other hand this file will usually change _with nearly every commit_. But maybe that's the only correct way...
* flypie.gresource: This should actually be re-created whenever one of the files mentioned in the xml changes.
* Clean looks clean to me.

**Comment by daPhipz** [comment]
> The *.po files are updated by weblate

OK, so we mustn't touch them at all? Or can we update them on the repo side and Weblate will pull the changes?

> The *.pot file will now be updated whenever `make` is executed?

This is not the case - have a look at this code snippet of the `Makefile`:

```Makefile
[…]
po/flypie.pot: $(JS_FILES) $(UI_FILES)
	@echo "Generating 'flypie.pot'..."
	@xgettext --from-code=UTF-8 \
        	  --add-comments=Translators \
        	  --copyright-holder="Simon Schneegans" \
        	  --package-name="Fly-Pie" \
        	  --output=po/flypie.pot \
        	  $(JS_FILES) $(UI_FILES)
[…]
```

The **target** (`po/flypie.pot`) will be regenerated when its **prerequisites** (or their prerequisites etc - this is recursive) are newer than the target. So whenever any `js` or `ui` file changed, the `pot` file gets updated. This is, of course, fairly often the case.

We could decouple updating the translation files from zip creation completely and maybe move that to a `PHONY` target called `translations` - which would de-clutter the git working directory, but also mean that you would have to run two commands (`make translations` and `make build`) in order to get a zip with the newest translation updates.

How about adding up-to-date translations as a prerequisite to the `release` target, but not the `build` target? That way, we can ignore translations when you just want to test something and need a zip, but when a release needs to be built, the translations get updated as well. This sounds like a good compromise to me - what are your thoughts?

**Comment by Schneegans** [comment]
I think we do not need an updated `pot` file for the release, right? It's not even included in the release zip. We also do not need to update the `po` files either, I guess. The main problem with updating the `pot` automatically is that it will lead to extreme merge conflicts as this file will be changed in each branch. So I think along these lines:

* `pot` file should _only_ be updated manually and _only_ in the `develop` branch. I'll do this frequently.
* `po` files are managed by weblate. Usually we shouldn't touch them. We can consider supporting manual translation, but then we will need to check whether changes are properly integrated into weblate - to be honest, I haven't tested this yet!
* `mo` files should be created on `make build`

What do you think?

**Comment by daPhipz** [comment]
Sounds reasonable! I'll make the `pot` a target then, but not incorporate it in the actual extension build as a dependency. That way, you can still `make pot` manually from time to time, but it won't mess up anything.

**Comment by daPhipz** [comment]
From my side, I think everything is ready to merge now. I am pretty confident that I didn't miss anything in the docs, but you may want to have a look nevertheless xD

File: docs/translating.md [hunk 2]
```diff
@@ -43,26 +45,41 @@ A pull request including your changes will be created automatically.
...
+´´´
+
+Then restart GNOME Shell with <kbd>Alt</kbd> + <kbd>F2</kbd>, <kbd>r</kbd> + <kbd>Enter</kbd>.
 Or logout / login if you are on Wayland.
 
+### Start a translation to a new language
+
+If you want to contribute a translation to a completely new language, this is best done via Weblate (see above).
+Direct PRs to add a new translation to the Fly-Pie project are also supported.
+However, automatic creation of a completely new `po` file with `make` is in the works, but not yet implemented.
```
Comment (daPhipz) [inline docs/translating.md:62]: I will add this feature later, if you want. :tada:

**Comment by Schneegans** [comment]
Thank you very much! I played around and it seems to work nicely! I only found one missing dependency which got lost (maybe due to the removal of the `pot` in-between): b5332e418694b6fb4bab5c7e3308948a2b096e95

**Comment by Schneegans** [comment]
I also removed the documentation regarding manual translation completely for now. It was a bit incomplete in the current state. But I think we should reconsider this at some point!

**Pull request merged**